### PR TITLE
Updated continuous integration

### LINKED ISSUE #39: Compatibility with NumPy 2.0 and Python 3.12
NumPy 2.0.dev can be pulled from https://anaconda.org/scientific-python-nightly-wheels/numpy

Python 3.12 is available on Actions (https://github.com/actions/python-versions/releases)

### PR DESCRIPTION
**Pull request by astrofrog**
Just simplifying the CI and adding recent versions of pytest

Fix #39

### CODE REVIEW

**Comment by pllim** [comment]
Should we just drop `pytest<6`?

**Comment by pllim** [comment]
CI disabled. Gonna close/reopen to trigger it.

**Comment by pllim** [comment]
You wanna be adventurous and have pytestdev use Python 3.12? 😅

**Comment by astrofrog** [comment]
Does Python 3.12 work with the openastronomy workflows?

**Comment by pllim** [comment]
I have no idea. I just know @bsipocz added a Python 3.12 job for pytest-doctestplus

**Comment by astrofrog** [comment]
@pllim lets just merge that for now if you are happy with it and I can investigate separately?

File: .github/workflows/ci_workflows.yml
Comment (astrofrog) [reply .github/workflows/ci_workflows.yml:28]: The wheels are always built in PRs. It's so fast there's no point making it optional.

Comment (astrofrog) [reply .github/workflows/ci_workflows.yml:28]: Wheels will always be built but only published to PyPI on a tag

Comment (pllim) [reply .github/workflows/ci_workflows.yml:28]: OK so after we have merged this, we should delete that "Build wheels" label.

**Comment by bsipocz** [comment]
IMO, do one last release before dropping 3.7, that's what we did elsewhere. Or is it hopelessly incompatible already?

**Comment by bsipocz** [comment]
Re py3.12: there wasn't any big issues, but this one fix was needed, which may or may not effect this plugin: https://github.com/scientific-python/pytest-doctestplus/pull/215

**Comment by astrofrog** [comment]
Can add back python 3.7, not a problem!

**Comment by astrofrog** [comment]
I've added back Python 3.7, hopefully this should still pass

File: tox.ini
Comment (pllim) [inline tox.ini:3]:
```suggestion
    py{37,38,39,310,311}-test{,-devdeps}
```

**Review by pllim — APPROVED** [review]
LGTM but I'll let Brigitta have a look too. Thanks!

File: .github/workflows/ci_workflows.yml
Comment (bsipocz) [inline .github/workflows/ci_workflows.yml:12]: I don't necessarily see why relying on one more layer makes sense here as all the CI setup is super basic for the plugin.

But if you want to, sure.

File: .github/workflows/ci_workflows.yml
Comment (bsipocz) [inline .github/workflows/ci_workflows.yml:15]: Somewhat unrelated improvement: instead of `pytest46` I like to call it `pytestoldest`, and pin it to the actual stated minimum version rather than a wildcard in that series. E.g. ``pytest==4.6.0``

If a fix is needed from an actual bugfix, then bump the minimum version to reflect that.

(for the rest of the versions I feel the flexible wildcard is fine, we need a wide enough matrix to cover over versions but don't need to test everything).

**Review by bsipocz — COMMENTED** [review]
Versions certainly need to be updated, but I don't see why relying on a middle layer makes sense for these super simple plugins.

**Comment by bsipocz** [comment]
Oh, and I can repeat what @pllim said above, please add py3.12 testing (can be done separately, but it's really time to make stuff compatible if it's not yet)

**Comment by pllim** [comment]
FWIW, I opened:

* https://github.com/astropy/pytest-arraydiff/issues/39

File: .github/workflows/ci_workflows.yml
Comment (astrofrog) [reply .github/workflows/ci_workflows.yml:12]: It's a very well maintained middle layer though and does save quite a few lines. I would have a preference to use it if we can but happy to revert if you feel strongly about it.

File: .github/workflows/ci_workflows.yml
Comment (astrofrog) [reply .github/workflows/ci_workflows.yml:15]: Sounds good, will do

**Comment by astrofrog** [comment]
This fails on Python 3.12 - @pllim @bsipocz is that an error you've seen before?

File: .github/workflows/ci_workflows.yml
Comment (pllim) [inline .github/workflows/ci_workflows.yml:27]: You have to use unreleased numpy (and any unreleased thingies that depend on it) for Python 3.12.

File: tox.ini
Comment (pllim) [inline tox.ini:30]: This is redundant. You already use pytestdev with py312.

```suggestion
```

Comment (astrofrog) [reply tox.ini]: oops copy/paste issue :sweat_smile:

**Comment by pllim** [comment]
I don't think you can pull astropy or tables for py312. You also have to pull in pandas dev.

https://github.com/astropy/pytest-arraydiff/blob/e006ea19eac35dcb287396caa0f93a68d35de5bf/setup.cfg#L37-L41

**Comment by astrofrog** [comment]
Hmm at the moment astropy is a required dependency for the test suite. Rather than re-write everything maybe we can just wait?

**Comment by pllim** [comment]
Wait for a miracle? 

* https://github.com/astropy/astropy/pull/14784
* https://github.com/astropy/astropy/pull/15298

File: .github/workflows/ci_workflows.yml
Comment (bsipocz) [reply .github/workflows/ci_workflows.yml:12]: No strong feelings about it here, but I wouldn't want it for doctestplus and remotedata.

File: tox.ini
Comment (bsipocz) [inline tox.ini:14]: You can also use the 1.26 RC, too, it's on the normal pypi already.

Comment (pllim) [reply tox.ini]: Not an issue anymore. I cleaned this up in a follow-up commit.

**Comment by pllim** [comment]
Gonna merge and then do a release. 🤞

**Pull request merged**
